// Protobuf messages used to model digital transmitters used in communications.
//
// Copyright (c) Aalyria Technologies, Inc., and its affiliates.
// Confidential and Proprietary. All rights reserved.

syntax = "proto2";

package minkowski.proto;

import "api/common/channel.proto";
import "api/common/wireless.proto";

option java_package = "com.aalyria.minkowski.proto";
option go_package = "aalyria.com/minkowski/api/common";

// Models a set of transmit channels.
message TxChannels {
  message TxChannelParams {
    // The maximum possible transmit power on this channel.
    optional double max_power_watts = 1;
  }

  // Maps the channel's center frequency, in Hz, to channel parameters.
  // For RF transceivers, the map key is the carrier frequency.
  // For optical transceivers, the key may be converted to wavelength.
  map<uint64, TxChannelParams> channel = 1;
}

// Models a wireless transmitter.
message TransmitterDefinition {
  reserved 1, 3 to 12, 16, 17 to max;

  // An optional friendly, human-readable string.
  optional string name = 2;

  // Maps a band profile ID to a set of channels.
  map<string, TxChannels> channel_set = 14;

  // Determines whether or not the SDN performs channel assignment.
  optional ChannelSelectionMethod channel_selection_method = 13;

  // Allows the user to configure an ordered sequence that describes how the
  // wireless transmission is created or modified prior to the input of
  // the platform antenna that propagates the signal.
  // protolint:disable:next REPEATED_FIELD_NAMES_PLURALIZED
  repeated TransmitSignalProcessor signal_processing_step = 15;
}

// Defines how a wireless signal is created, modified, or propagated.
message TransmitSignalProcessor {
  oneof type {
    AmplifierDefinition amplifier = 1;
    MiscGainOrLoss gain_or_loss = 2;
  }
}
